### DISCUSSION TITLE
Cannot Use MySQL variables within Transaction

### DISCUSSION THREAD

**Discussion by marwan-at-work** [planetscale/discussion#22] [Bug Reports] [answered]
Hi there 👋🏼 
I am consistently able to reproduce a weird bug where if I use arguments in an INSERT query within a transaction, things succeed but the row is never written. Arguments outside of a transaction work just fine. 

If you cannot reproduce with the following, I have live planet scale database that I'm happy to do a live pairing with and show you the error. 

To reproduce: 

1. Create a planet scale beta database.
2. Create a branch with a "users" table that has a unique "email" column. And merge into main.
3. Run `pscale connect <db> main` to open a local proxy.
4. Run the following Go program:

```golang
package main

import (
	"context"
	"database/sql"

	// register driver
	_ "github.com/go-sql-driver/mysql"
)

var ctx = context.Background()

func main() {
	ctx := context.Background()
	db, err := db.Open("mysql", "root:@tcp(127.0.0.1:3306)/<db>?tls=false&parseTime=true") // make sure to replace db
	must(err)
	tx, err := db.BeginTx(ctx, &sql.TxOptions{})
	must(err)
	_, err = tx.Exec(`insert into users (email) values (?)`, "hello@world.com")
	must(err)
	must(tx.Commit())
}

func must(err error) {
	if err != nil {
		panic(err)
	}
}
```

5. From a shell or another Go program run: `select * from users;` 
6. Note that the row was never inserted. 

Note that the following alterations all work successfully (which is why I narrowed it down to args-within-transactions): 

```golang
// transaction, no args.
_, err = tx.Exec(`insert into users (email) values ('hello@world.com')`) // this works
```

```golang
// no transaction, with args
_, err = db.Exec(`insert into users (email) values (?)`, "hello@world.com") // this works
```

Thanks!

**Comment by aquarapid** [comment]
Confirmed;  something is off with our prepared statement handling via the pscale proxy.  We are looking into it.  For the moment, as you mentioned, the workaround is to not use a prepared statement.

**Reply by marwan-at-work** [reply]
@aquarapid unfortunately that's an impossible workaround since it would make the application vulnerable to sql injections (unless I'm missing a way where you can avoid prepared statements _and_ avoid sql injection). Looking forward for the fix and thanks for confirming! 🎉

**Reply by enisoc** [reply]
If you enable [interpolateParams](https://github.com/go-sql-driver/mysql#interpolateparams) like below, the client will do the interpolation for you in a safe way before sending the query to the server:

```go
db, err := db.Open("mysql", "root:@tcp(127.0.0.1:3306)/<db>?tls=false&parseTime=true&interpolateParams=true")
```

**Reply by shlomi-noach** [reply]
:100: to the above. As a `golang` developer I found `interpolateParams` to work well, plus it saves a round trip to the database so it performs better.

**Reply by Ankita2809** [reply]
While connecting mysql with netbeans, i found error showing jdbc driver..it doesnt show connection on local host..help me to find solution

**Reply by shlomi-noach** [reply]
@Ankita2809 I'm not sure if your question is related to PlanetScale? If so, can you please open a new discussion? Thank you :pray:

**Comment by mattrobenolt** [comment, marked as answer]
Just for status update, we've 100% reproduced and are tracking down a fix. Oddly enough if you run a non-prepared statement first within a transaction, then other prepared statements after that are fine. The issue only pops up if the first statement within the transaction is prepared.

Thanks for pointing this out!

**Reply by mattrobenolt** [reply]
This has been fixed and verified in production.

**Reply by marwan-at-work** [reply]
Can confirm, thank you 🎉
